akka-stream-tests/src/test/scala/akka/stream/scaladsl/RetryFlowSpec.scala

```diff
@@ -0,0 +1,406 @@
+        .viaMat(
+          RetryFlow.withBackoff(
+            8,
+            10.millis,
+            5.second,
+            0,
+            alwaysFailingFlow
+              .via(Utils.delayCancellation(10.seconds))
+              .viaMat(KillSwitches.single[(Try[Int], Int)])(Keep.right))(alwaysRecoveringFunc))(Keep.both)
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+      source.sendNext(1)
+      sink.expectNoMessage()
+      killSwitch.abort(failedElem.failed.get)
+      sink.expectError(failedElem.failed.get)
+    }
+
+    "tolerate killswitch abort after the RetryFlow while on retry spin" in {
+      val ((source, killSwitch), sink) = TestSource
+        .probe[Int]
+        .map(i => (i, i))
+        .via(stuckForeverRetrying)
+        .via(Utils.delayCancellation(10.seconds))
+        .viaMat(KillSwitches.single[(Try[Int], Int)])(Keep.both)
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+
+      source.sendNext(1)
+      sink.expectNoMessage()
+
+      killSwitch.abort(failedElem.failed.get)
+      sink.expectError(failedElem.failed.get)
+    }
+
+    "finish only after processing all elements in stream" in {
+      val (source, sink) = TestSource
+        .probe[Int]
+        .map(i => (i, i))
+        .via(RetryFlow.withBackoff(8, 10.millis, 5.second, 0, flow[Int]) {
+          case (Failure(_), x) => Some(List.fill(x)(1 -> 1))
+        })
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+
+      source.sendNext(1)
+      sink.expectNext((Success(2), 1))
+
+      source.sendNext(3)
+      sink.expectNext((Success(4), 3))
+
+      source.sendNext(2)
+      source.sendComplete()
+      sink.expectNext((Success(2), 1))
+      sink.expectNext((Success(2), 1))
+
+      sink.expectComplete()
+    }
+
+    "exponentially backoff between retries" in {
+      val NumRetries = 7
+
+      case class State(counter: Int, retriedAt: List[Long])
+
+      val flow = Flow.fromFunction[(Int, State), (Try[Int], State)] {
+        case (i, j) if i % NumRetries == 0  => (Success(i), j)
+        case (_, State(counter, retriedAt)) => (failedElem, State(counter + 1, System.currentTimeMillis() :: retriedAt))
+      }
+
+      val (source, sink) = TestSource
+        .probe[Int]
+        .map(i => (i, State(i, Nil)))
+        .via(RetryFlow.withBackoff(8, 10.millis, 5.second, 0, flow) {
+          case (Failure(_), s @ State(counter, _)) => Some(List(counter -> s))
+        })
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(1)
+
+      source.sendNext(1)
+      val (result, State(_, retriedAt)) = sink.expectNext()
+
+      result shouldBe Success(NumRetries)
+      val timesBetweenRetries = retriedAt
+        .sliding(2)
+        .collect {
+          case before :: after :: Nil => before - after
+        }
+        .toIndexedSeq
+
+      timesBetweenRetries.reverse should strictlyIncrease
```

Comment:
can we use `System.nanoTime` in this test, since we are measuring durations?

Reply:
here is another `currentTimeMillis`, this should be easy to replace with nanoTime